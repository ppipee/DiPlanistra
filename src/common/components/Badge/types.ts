export enum BadgeVariant {
	Outlined = 'outlined',
	Filled = 'filled',
}

export type BadgeVariantTypes = BadgeVariant | 'filled' | 'outlined'

export interface BadgeProps {
	$variant?: BadgeVariantTypes
	$color: string
	$secondColor?: string
}
