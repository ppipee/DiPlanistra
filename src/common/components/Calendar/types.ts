export interface CalenderDateProps {
	startDate?: Date
	endDate?: Date
}

export interface CalendarDisplayStateProps {
	year: number
	month: number
}

export interface CalendarDisplayRange {
	startRange?: CalendarDisplayStateProps
	endRange?: CalendarDisplayStateProps
}

export interface CalendarCellProps {
	$inBetween?: boolean
	$selected?: boolean
	$variant?: CalendarCellVariant
	$isFirstCell?: boolean
	$isLastCell?: boolean
	$isDateSelectable?: boolean
}

export type CalendarCellVariant = 'currentMonth' | 'otherMonth' | 'currentDate'
