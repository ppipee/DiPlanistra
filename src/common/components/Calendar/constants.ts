import { main } from 'common/styles/colors'

export const DAYS = ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa']

export const CALENDAR_COLOR = {
	Active: main[500],
	Hover: '#7CC0C6',
	Between: '#BFE7DF',
}
