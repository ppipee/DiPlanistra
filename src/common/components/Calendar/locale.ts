export const CALENDAR_START_DATE_TITLE = ['เริ่มต้น', 'Start Date']
export const CALENDAR_END_DATE_TITLE = ['สิ้นสุด', 'End Date']
