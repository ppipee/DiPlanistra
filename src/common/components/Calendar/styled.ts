import styled from 'styled-components'

import spaces from 'common/styles/mixins/spaces'

export const CalendarContainer = styled.div`
	padding: ${spaces(4)};
	display: inline-block;
	width: 280px;
	user-select: none;
`
