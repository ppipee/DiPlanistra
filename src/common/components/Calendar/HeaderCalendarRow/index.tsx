import React, { memo } from 'react'

import { Cell, CellWrapper } from '../components/CalendarCell/styled'
import { RowContainer } from '../components/CalendarRow/styled'
import { DAYS } from '../constants'

const HeaderCalendarRow = () => {
	return (
		<RowContainer>
			{DAYS.map((day) => (
				<CellWrapper key={`header-${day}`}>
					<Cell className="heading4">{day}</Cell>
				</CellWrapper>
			))}
		</RowContainer>
	)
}

export default memo(HeaderCalendarRow)
