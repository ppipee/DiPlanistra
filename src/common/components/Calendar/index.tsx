import React, { ReactNode } from 'react'

import { CalendarDisplayRange, CalenderDateProps } from 'common/components/Calendar/types'

import { CalendarCellDate } from './components/CalendarCell'
import CalendarHeader from './components/CalendarHeader'
import CalendarProvider from './components/CalendarProvider'
import CalendarTable from './components/CalendarTable'
import { CalendarContainer } from './styled'

export type Props = {
	defaultDate?: CalenderDateProps
	setDate: (date: CalenderDateProps) => void
	isDateSelectable?: (date: CalendarCellDate) => boolean
	dateRangeShown?: CalendarDisplayRange
	children?: ReactNode
}

const Calendar = ({ defaultDate, setDate, isDateSelectable, dateRangeShown, children }: Props) => {
	return (
		<CalendarProvider
			isDateSelectable={isDateSelectable}
			defaultDate={defaultDate}
			setDate={setDate}
			dateRangeShown={dateRangeShown}
		>
			<CalendarContainer>
				<CalendarHeader />
				<CalendarTable />
				{children}
			</CalendarContainer>
		</CalendarProvider>
	)
}

export default Calendar
