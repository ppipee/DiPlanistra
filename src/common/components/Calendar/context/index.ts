import { createContext } from 'react'

import { CalendarCellDate } from '../components/CalendarCell'
import { CalenderDateProps, CalendarDisplayStateProps, CalendarDisplayRange } from '../types'

export interface CalendarContextType {
	date: CalenderDateProps
	setDate: (date: CalenderDateProps) => void
	displayState: CalendarDisplayStateProps
	setDisplayState: (state: CalendarDisplayStateProps) => void
	dateHover?: Date
	setDateHover: (date?: Date) => void
	isDateSelectable: (date: CalendarCellDate) => boolean
	dateRangeShown?: CalendarDisplayRange
}

const CalendarContext = createContext<CalendarContextType | null>(null)

export default CalendarContext
