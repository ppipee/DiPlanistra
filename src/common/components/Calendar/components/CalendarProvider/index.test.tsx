import React from 'react'

import { render } from '@testing-library/react'
import 'jest-styled-components'

import { CalenderDateProps } from 'common/components/Calendar/types'

import * as CalendarProviderType from '.'

describe('Test CalendarProvider', () => {
	const CalendarContextSpy = jest.fn()

	jest.doMock('common/components/Calendar/context', () => ({
		Provider: (props: Record<string, any>) => {
			CalendarContextSpy(props)

			return <div {...props}>{props.children}</div>
		},
	}))

	const { default: CalendarProvider } = require('.') as typeof CalendarProviderType
	const setDate = (date: CalenderDateProps) => {}

	it('should render CalendarProvider correctly', () => {
		const { container } = render(
			<CalendarProvider setDate={setDate}>
				<div />
			</CalendarProvider>,
		)

		expect(container.firstChild).toMatchSnapshot()
	})

	it('should render CalendarProvider with defaultDate', () => {
		const { container } = render(
			<CalendarProvider
				setDate={setDate}
				defaultDate={{ startDate: new Date('1998/09/07'), endDate: new Date('1998/09/07') }}
			>
				<div />
			</CalendarProvider>,
		)

		expect(container.firstChild).toMatchSnapshot()
	})

	it('should render CalendarProvider with defaultDate but not container endDate', () => {
		const { container } = render(
			<CalendarProvider setDate={setDate} defaultDate={{ startDate: new Date('1998/09/07') }}>
				<div />
			</CalendarProvider>,
		)

		expect(container.firstChild).toMatchSnapshot()
	})
})
