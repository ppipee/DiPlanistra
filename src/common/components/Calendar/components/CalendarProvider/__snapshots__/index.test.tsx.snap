// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CalendarProvider should render CalendarProvider correctly 1`] = `
<div
  value="[object Object]"
>
  <div />
</div>
`;

exports[`Test CalendarProvider should render CalendarProvider with defaultDate 1`] = `
<div
  value="[object Object]"
>
  <div />
</div>
`;

exports[`Test CalendarProvider should render CalendarProvider with defaultDate but not container endDate 1`] = `
<div
  value="[object Object]"
>
  <div />
</div>
`;
